From 64aaf17e26edff748ea8560773ea9a177cddce50 Mon Sep 17 00:00:00 2001
From: Macross Chen <macross.chen@mediatek.com>
Date: Tue, 28 Dec 2021 16:21:33 +0800
Subject: [PATCH] v4l2videoenc: set caps with prepend header flag

query prepend header from driver and set to output caps.

Signed-off-by: Macross Chen <macross.chen@mediatek.com>
---
 sys/v4l2/gstv4l2videoenc.c | 26 ++++++++++++++++++++++++++
 1 file changed, 26 insertions(+)

diff --git a/sys/v4l2/gstv4l2videoenc.c b/sys/v4l2/gstv4l2videoenc.c
index d43f649..11505a8 100644
--- a/sys/v4l2/gstv4l2videoenc.c
+++ b/sys/v4l2/gstv4l2videoenc.c
@@ -301,6 +301,29 @@ done:
   return ret;
 }
 
+static void
+gst_v4l2_video_enc_set_driver_caps (GstVideoEncoder * encoder,
+    GstVideoCodecState * state)
+{
+  GstV4l2VideoEnc *self = GST_V4L2_VIDEO_ENC (encoder);
+  GstV4l2Object *v4l2object = self->v4l2output;
+  struct v4l2_control control = { 0, };
+
+  state->caps = gst_caps_make_writable (state->caps);
+
+  /* check prepend header */
+  control.id = V4L2_CID_MPEG_VIDEO_PREPEND_SPSPPS_TO_IDR;
+  if (v4l2object->ioctl (v4l2object->video_fd, VIDIOC_G_CTRL, &control) < 0) {
+    GST_DEBUG_OBJECT (v4l2object->dbg_obj,
+        "Get prepend_hdr fail, set to default 0.");
+    control.value = 0;
+  } else {
+    GST_DEBUG_OBJECT (v4l2object->dbg_obj, "prepend_hdr = %d", control.value);
+  }
+  gst_caps_set_simple (state->caps, "prepend_hdr", G_TYPE_INT, control.value,
+      NULL);
+}
+
 static gboolean
 gst_v4l2_video_enc_set_format (GstVideoEncoder * encoder,
     GstVideoCodecState * state)
@@ -329,6 +352,9 @@ gst_v4l2_video_enc_set_format (GstVideoEncoder * encoder,
     self->input_state = NULL;
   }
 
+  /* query & set caps from driver */
+  gst_v4l2_video_enc_set_driver_caps (encoder, state);
+
   outcaps = gst_pad_get_pad_template_caps (encoder->srcpad);
   outcaps = gst_caps_make_writable (outcaps);
   output = gst_video_encoder_set_output_state (encoder, outcaps, state);
